lc-compile: Print contextual information for each warning or error

peter-b opened the pull request:
This patch modifies **lc-compile**'s warning and error output to include the affected line and an indication of where in the line the problem occurs.  For example, when compiling the following module:

```
module org.example.errors
constant kBadConstant is InvalidExpression
end module
```

lc-compile will now output:

```
foo.lcb:2:26: error: Identifier 'InvalidExpression' not declared
 constant kBadConstant is InvalidExpression
                          ^
```

This is implemented by adding a lazily-constructed array of line strings to the `File` structure in **lc-compile**. The position data provided by **gentle** doesn't include a file offset, and scanning the whole file for each position to find the line in which it occurs would be really expensive.  Instead, **lc-compile** will now compute and cache an array of every line in a source file the first time that a position's source file context is requested.

peter-b committed:
lc-compile: Add general-purpose debugging message function

Add a `Debug()` function to generate debugging output that isn't
related to bytecode emission or to dependency calculation.

peter-b committed:
lc-compile: Clean up some unused variables in _WarningS().

peter-b committed:
lc-compile: Remove the unused stream member from the File structure.

This isn't actually used for anything, and its presence is a bit
confusing.

peter-b committed:
lc-compile: Add lazily-constructed array of file lines to File structure

For printing out context of warnings and errors generated by
lc-compile, it's necessary to be able to get access of the text of
each line in the source file.

The position doesn't include a file offset, and scanning the whole
file for each position to find its surrounding context would be
excessively expensive.  Instead, this patch takes the approach of
loading the file's text into a buffer, and then creating an array of
pointers into the buffer that point to the start of each line.

When a file doesn't have any warnings or errors, there's no point in
going to the trouble of setting it up, so the file is scanned lazily
the first time that a particular line's text is needed.

peter-b committed:
lc-compile: Print contextual information for each warning or error

This patch modifies lc-compile's warning and error output to include
the affected line and an indication of where in the line the problem
occurs.  For example, when compiling the following module:

    module org.example.errors
    constant kBadConstant is InvalidExpression
    end module

lc-compile will now output:

    foo.lcb:2:26: error: Identifier 'InvalidExpression' not declared
     constant kBadConstant is InvalidExpression
                              ^

peter-b committed:
lc-compile: Add release note for warning message improvements

livecodefraser commented:
Just need some colourisation and wavy lines to indicate ranges and and and a pony :wink:

@livecode-vulcan review okay 5c7170a

livecode-vulcan commented:
💙 review by @livecodefraser ok 5c7170a3338b96c7cb74a3b86b5dc499b0d753fe

livecode-vulcan commented:
😎 test success 5c7170a3338b96c7cb74a3b86b5dc499b0d753fe
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/231)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/235)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/231)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/231)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/235)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/231)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/231)

montegoulding commented on toolchain/lc-compile/src/report.c:
```diff
@@ -172,6 +205,7 @@ static void _Warning(long p_position, const char *p_message)
     {
         _PrintPosition(p_position);
         fprintf(stderr, "warning: %s\n", p_message);
+        _PrintContext(p_position);
```
@peter-b I'm wondering why `_Warning`, `_ErrorS` and `_WarningS` all include the context but `_Error` doesn't.

peter-b replied:
@montegoulding Thanks, I've fixed that.

montegoulding commented on toolchain/lc-compile/src/report.c:
```diff
@@ -152,6 +169,22 @@ static void _PrintPosition(long p_position)
+/* Print the source code line that contains p_position, and another
+ * line below it with a caret pointing "up" to the exact character
+ * that the position specifies. */
+static void _PrintContext(long p_position)
```
@peter-b Could this output either be enabled or skipped by an option. It wouldn't be needed by anything parsing the output and displaying the error itself. Perhaps `-v` would be good...

peter-b replied:
@montegoulding IMHO it should be the default output format, just like pretty much every other compiler out there.

At the moment we don't have anything that parses the output that could get confused by this.  In the long term, when we actually have a tool that can't we can probably add a designed-to-be-machine-readable output format for messages (for example, @runrevmark has suggested adding a `--json` mode where _all_ output is JSON on stdout).

Note that all "context" output is guaranteed to be on lines starting with whitespace.

montegoulding reviewed (commented):
Very handy

peter-b committed:
lc-compile: Make sure source context is printed for all errors

montegoulding commented:
@livecode-vulcan review ok a7950bf

livecode-vulcan commented:
💙 review by @montegoulding ok a7950bf91e140fe7440837061dc5f1255edba792

livecode-vulcan commented:
😎 test success a7950bf91e140fe7440837061dc5f1255edba792
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/239)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/243)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/239)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/239)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/243)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/239)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/239)

peter-b committed:
Make compiler front-end tests check warning/error context info.

Ensure that the tests that check for compiler warnings/errors also
check that the corresponding source code line and position marker
are emitted in the output of lc-compile.

livecodefraser commented:
@livecode-vulcan review okay 5148f95

livecode-vulcan commented:
💙 review by @livecodefraser ok 5148f951dcbb33a53e9fe36a930376298eccb69a

livecode-vulcan commented:
😎 test success 5148f951dcbb33a53e9fe36a930376298eccb69a
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/244)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/248)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/244)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/244)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/248)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/244)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/244)
